Model/MetaDataModel.php

```diff
@@ -0,0 +1,184 @@
+<?php
+/**
+ * Shop System Plugins:
+ * - Terms of Use can be found under:
+ * https://github.com/wirecard/oxid-ee/blob/master/_TERMS_OF_USE
+ * - License can be found under:
+ * https://github.com/wirecard/oxid-ee/blob/master/LICENSE
+ */
+
+namespace Wirecard\Oxid\Model;
+
+use OxidEsales\Eshop\Core\DatabaseProvider;
+use OxidEsales\Eshop\Core\Field;
+use OxidEsales\Eshop\Core\Registry;
+
+use OxidEsales\EshopCommunity\Core\Model\BaseModel;
+
+/**
+ * Adds meta data functionality to a model.
+ *
+ * @since 1.2.0
+ */
+trait MetaDataModel
+{
+    /**
+     * Returns the name of the meta data table for this model.
+     * @return string
+     *
+     * @since 1.2.0
+     */
+    abstract public function getTableName();
+
+    /**
+     * Returns the short field name for a given field (e.g. oxpayments__foo → foo).
+     *
+     * @param string $sLongFieldName
+     * @return string
+     *
+     * @since 1.2.0
+     */
+    protected function _getFieldShortName($sLongFieldName)
+    {
+        return str_replace("{$this->_sCoreTable}__", '', $sLongFieldName);
+    }
+
+    /**
+     * Extends the `load` method to also load meta fields from the database.
+     *
+     * @inheritdoc
+     * @param string $sOxid
+     * @return bool
+     *
+     * @since 1.2.0
+     */
+    public function load($sOxid)
+    {
+        $bIsLoaded = is_subclass_of($this, BaseModel::class) ? parent::load($sOxid) : true;
+
+        foreach ($this->loadMetaData() as $sFieldShortName => $mFieldValue) {
+            $sFieldLongName = $this->_getFieldLongName($sFieldShortName);
+
+            $this->$sFieldLongName = new Field($mFieldValue);
+        }
+
+        return $bIsLoaded;
+    }
+
+    /**
+     * Extends the `save` method to also save meta fields to the database.
+     *
+     * @inheritdoc
+     * @return string|bool
+     *
+     * @since 1.2.0
+     */
+    public function save()
+    {
+        $aMetaData = [];
+
+        foreach ($this->getMetaFields() as $sFieldLongName => $oField) {
+            $sFieldShortName = $this->_getFieldShortName($sFieldLongName);
+
+            if ($this->$sFieldLongName instanceof Field) {
+                $aMetaData[$sFieldShortName] = $this->$sFieldLongName->getRawValue();
+            }
+        }
+
+        $this->deleteMetaData();
```

Comment:
On every save you delete the all setting before you save them again?

Reply:
Correct. This is needed because there might be obsolete keys that should be deleted upon re-saving.

Reply:
It is not efficient to delete all keys and then add new ones when you do for example an update of the config in the backend. Is it not possible to add an update or why don't you update values incited of deleting and saving them?

Reply:
The reason for this was that stray entries in the database could be created without noticing. Consider this example:

```php
$oPayment->foo = new Field('foo');
$oPayment->bar = new Field('bar');
$oPayment->save();

// now, both `foo` and `bar` are saved to the database, however:

$oPayment->baz = new Field('baz');
$oPayment->save();

// now `baz` is saved, but `foo` and `bar` are still in the database, even though they were not deliberately added
```

I decided to turn it around now, so that in order to _delete_ a field from an object, you have to specifically set it to `null`, e.g. given the above example:

```php
$oPayment->foo = null;
$oPayment->bar = null;
$oPayment->save();

// now `foo` and `bar` are deleted and only `baz` remains
```

Reply:
Great